RockSniffer/RPC/AlbumArtResolver.cs

```diff
@@ -0,0 +1,136 @@
+﻿using Newtonsoft.Json;
+using RockSnifferLib.Logging;
+using RockSnifferLib.Sniffing;
+using System;
+using System.Collections.Generic;
+using System.Collections.Specialized;
+using System.IO;
+using System.Net;
+using System.Net.Http;
+
+namespace RockSniffer.RPC
+{
+    public class AlbumArtResolver
+    {
+        [Serializable]
+        private class AppleResult
+        {
+            public string artistName;
+            public string collectionName;
+            public string trackName;
+            public string artworkUrl100;
+        }
+
+        [Serializable]
+        private class AppleRequestResponse
+        {
+            public int resultCount;
+            public AppleResult[] results;
+        }
+
+        private HttpClient httpClient = new HttpClient();
+        private Dictionary<string, (string URL, string DisplayText)> cache = new Dictionary<string, (string URL, string DisplayText)>();
+
+
+        public AlbumArtResolver() {
+            httpClient.DefaultRequestHeaders.Add("User-Agent", "RockSniffer");
+        }
+
+        /// <summary>
+        /// Attempt to obtain album cover
+        /// </summary>
+        /// <param name="songInfo">Details about the given song</param>
+        public (string URL, string DisplayText)? Get(SongDetails songInfo)
+        {
+            string key = $"{songInfo.artistName}|{songInfo.albumName}";
+
+            // Use double-checked locking pattern, since there are multiple threads
+            if (!cache.ContainsKey(key))
+                lock (this)
```

Comment:
It's just occurred to me to me that this is happening directly from the event, which is called each time memory is read, if web requests start backing up, i.e. multiple threads are waiting for this lock, does this hinder memory reading and other sniffer functions?

Reply:
I guess it technically might, yes. How often is the memory read, and does it have to be read every single time without any exceptions?

Reply:
Memory is read every 100ms, having extremely long delays (>500ms) might be noticeable on some addons as stuttering.
Do you think it would be possible / worth transitioning the album art fetching to explicitly the `OnStateChanged` event, to only occur when state changes to indicate a song is playing?

Reply:
I think the most robust solution could be to have a separate thread for updates of the Discord status. I would be willing to implement that, but I'm not sure if it's not too many changes to push in one merge request.